File: src/40_combination_sum_ii/CombinationSumII.java
import java.util.*;

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(0, new ArrayList<>(), 0, candidates, target, res);
        return res;
    }

    private void backtrack(int start, List<Integer> curr, int total, int[] candidates, int target,
            List<List<Integer>> res) {
        if (total == target) {
            res.add(new ArrayList<>(curr));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i] == candidates[i - 1])
                continue;
            if (total + candidates[i] > target)
                break;
            curr.add(candidates[i]);
            backtrack(i + 1, curr, total + candidates[i], candidates, target, res);
            curr.remove(curr.size() - 1);
        }
    }
}

public class CombinationSumII {
    private static String formatResult(List<List<Integer>> result) {
        List<String> sorted = new ArrayList<>();
        for (List<Integer> comb : result) {
            Collections.sort(comb);
            sorted.add(comb.toString());
        }
        Collections.sort(sorted);
        return sorted.toString();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        // Test case 1
        int[] c1 = { 10, 1, 2, 7, 6, 1, 5 };
        int t1 = 8;
        List<List<Integer>> e1 = Arrays.asList(Arrays.asList(1, 1, 6), Arrays.asList(1, 2, 5), Arrays.asList(1, 7),
                Arrays.asList(2, 6));
        System.out.println(formatResult(sol.combinationSum2(c1, t1)) + " | " + formatResult(e1));
        // Test case 2
        int[] c2 = { 2, 5, 2, 1, 2 };
        int t2 = 5;
        List<List<Integer>> e2 = Arrays.asList(Arrays.asList(1, 2, 2), Arrays.asList(5));
        System.out.println(formatResult(sol.combinationSum2(c2, t2)) + " | " + formatResult(e2));
        // Test case 3
        int[] c3 = { 2 };
        int t3 = 1;
        List<List<Integer>> e3 = Arrays.asList();
        System.out.println(formatResult(sol.combinationSum2(c3, t3)) + " | " + formatResult(e3));
        // Test case 4
        int[] c4 = { 1 };
        int t4 = 1;
        List<List<Integer>> e4 = Arrays.asList(Arrays.asList(1));
        System.out.println(formatResult(sol.combinationSum2(c4, t4)) + " | " + formatResult(e4));
        // Test case 5
        int[] c5 = { 1 };
        int t5 = 2;
        List<List<Integer>> e5 = Arrays.asList();
        System.out.println(formatResult(sol.combinationSum2(c5, t5)) + " | " + formatResult(e5));
    }
}

File: src/40_combination_sum_ii/combination_sum_ii.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> curr;
        backtrack(0, curr, 0, candidates, target, res);
        return res;
    }
    void backtrack(int start, vector<int>& curr, int total, vector<int>& candidates, int target,
                   vector<vector<int>>& res) {
        if (total == target) {
            res.push_back(curr);
            return;
        }
        for (int i = start; i < candidates.size(); i++) {
            if (i > start && candidates[i] == candidates[i - 1]) continue;
            if (total + candidates[i] > target) break;
            curr.push_back(candidates[i]);
            backtrack(i + 1, curr, total + candidates[i], candidates, target, res);
            curr.pop_back();
        }
    }
};

string formatResult(vector<vector<int>> result) {
    vector<string> sorted;
    for (auto& comb : result) {
        sort(comb.begin(), comb.end());
        string s = "[";
        for (size_t i = 0; i < comb.size(); i++) {
            s += to_string(comb[i]);
            if (i < comb.size() - 1) s += ",";
        }
        s += "]";
        sorted.push_back(s);
    }
    sort(sorted.begin(), sorted.end());
    string res = "[";
    for (size_t i = 0; i < sorted.size(); i++) {
        res += sorted[i];
        if (i < sorted.size() - 1) res += ",";
    }
    res += "]";
    return res;
}

int main() {
    Solution sol;
    // Test case 1
    vector<int> c1 = {10, 1, 2, 7, 6, 1, 5};
    int t1 = 8;
    vector<vector<int>> e1 = {{1, 1, 6}, {1, 2, 5}, {1, 7}, {2, 6}};
    cout << formatResult(sol.combinationSum2(c1, t1)) << " | " << formatResult(e1) << endl;
    // Test case 2
    vector<int> c2 = {2, 5, 2, 1, 2};
    int t2 = 5;
    vector<vector<int>> e2 = {{1, 2, 2}, {5}};
    cout << formatResult(sol.combinationSum2(c2, t2)) << " | " << formatResult(e2) << endl;
    // Test case 3
    vector<int> c3 = {2};
    int t3 = 1;
    vector<vector<int>> e3 = {};
    cout << formatResult(sol.combinationSum2(c3, t3)) << " | " << formatResult(e3) << endl;
    // Test case 4
    vector<int> c4 = {1};
    int t4 = 1;
    vector<vector<int>> e4 = {{1}};
    cout << formatResult(sol.combinationSum2(c4, t4)) << " | " << formatResult(e4) << endl;
    // Test case 5
    vector<int> c5 = {1};
    int t5 = 2;
    vector<vector<int>> e5 = {};
    cout << formatResult(sol.combinationSum2(c5, t5)) << " | " << formatResult(e5) << endl;
    return 0;
}
